synapse/visibility.py

```diff
@@ -92,6 +92,12 @@ def filter_events_for_client(store, user_id, events, is_peeking=False,
 
     erased_senders = yield store.are_users_erased((e.sender for e in events))
 
+    room_ids = list(e.room_id for e in events)
```

Comment:
Let's only do each room once...

```suggestion
    room_ids = set(e.room_id for e in events)
```